hazelcast/src/main/java/com/hazelcast/sql/impl/exec/scan/index/MapIndexScanExecIterator.java

```diff
@@ -0,0 +1,203 @@
+ */
+
+package com.hazelcast.sql.impl.exec.scan.index;
+
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.util.collection.PartitionIdSet;
+import com.hazelcast.map.impl.MapContainer;
+import com.hazelcast.query.impl.InternalIndex;
+import com.hazelcast.query.impl.QueryableEntry;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.exec.scan.KeyValueIterator;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.schema.map.MapTableUtils;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Iterator for index-based partitioned map access.
+ */
+@SuppressWarnings("rawtypes")
+public class MapIndexScanExecIterator implements KeyValueIterator {
+
+    private final Iterator<QueryableEntry> iterator;
+
+    private Data currentKey;
+    private Object currentValue;
+    private Data nextKey;
+    private Object nextValue;
+
+    public MapIndexScanExecIterator(
+        MapContainer map,
+        String indexName,
+        int expectedComponentCount,
+        IndexFilter indexFilter,
+        List<QueryDataType> expectedConverterTypes,
+        PartitionIdSet expectedPartitions,
+        ExpressionEvalContext evalContext
+    ) {
+        iterator = getIndexEntries(
+            map,
+            indexName,
+            indexFilter,
+            evalContext,
+            expectedComponentCount,
+            expectedConverterTypes,
+            expectedPartitions
+        );
+
+        advance0();
+    }
+
+    @Override
+    public boolean tryAdvance() {
+        if (!done()) {
+            currentKey = nextKey;
+            currentValue = nextValue;
+
+            advance0();
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean done() {
+        return nextKey == null;
+    }
+
+    @Override
+    public Object getKey() {
+        return currentKey;
+    }
+
+    @Override
+    public Object getValue() {
+        return currentValue;
+    }
+
+    private void advance0() {
+        if (iterator.hasNext()) {
+            QueryableEntry<?, ?> entry = iterator.next();
+
+            nextKey = entry.getKeyData();
+            nextValue = entry.getValue();
+        } else {
+            nextKey = null;
+            nextValue = null;
+        }
+    }
+
+    private Iterator<QueryableEntry> getIndexEntries(
+        MapContainer map,
+        String indexName,
+        IndexFilter indexFilter,
+        ExpressionEvalContext evalContext,
+        int expectedComponentCount,
+        List<QueryDataType> expectedConverterTypes,
+        PartitionIdSet expectedPartitions
+    ) {
+        // Find the index
+        InternalIndex index = map.getIndexes().getIndex(indexName);
+
+        if (index == null) {
+            throw QueryException.error(
```

Comment:
If the index has been removed concurrently, shall we throw this exception to the user instead of falling back to the map scan?